### ISSUE TITLE
`Callable`s that don't return `None` should be compatible with `Callable[[<args>], None]`

### ISSUE THREAD

**Issue by KotlinIsland** [python/typing#924]
I think that any function should be compatible with `Callable[<args>, None]` if it's arguments match, it's typesafe and many other languages behave that way.
```py
from typing import Callable

fn: Callable[[int], None] = lambda arg: arg  # currently an error
print(fn(1))  # error, function does not return a value
```

TypeScript:
```ts
const fn: (arg: number) => void = (arg) => arg  // no error
console.log(fn(1))
```

Kotlin:
```kt
val fn: (arg: Int) -> Unit = { it }  // no error
println(fn(1))
```

**Comment by TeamSpen210** [comment]
If you're ignoring the argument, you should probably define it as `Any`. Perhaps `object` actually, since then you'll get errors if you do anything with it.

**Comment by erictraut** [comment]
If you want to describe a function that can return anything, then `Callable[..., None]` is appropriate. The type `Callable[..., None]` describes a function that is guaranteed to return `None`. The type `None` is not equivalent to `void` in TypeScript.

**Comment by KotlinIsland** [comment]
@TeamSpen210 I don't want `object`, I want `void`. I explicitly want it to be unusable.
@erictraut I don't want `Callable[..., None]`. I explicitly want to pass it exactly `(int)`.

**Comment by TeamSpen210** [comment]
That's what `object` is, it has no methods and would fail to typecheck for most operations. The few it'll be allowed for would be things like `repr()` or equality checks, which would be valid to do.

**Comment by KotlinIsland** [comment]
> That's what `object` is, it has no methods and would fail to typecheck for most operations. The few it'll be allowed for would be things like `repr()` or equality checks, which would be valid to do.

`object` is quite distinct from `void`, which represents something that should not be used for anything, eg:

```py
def foo() -> None:  ...  # a None return type means void

print(f"Hi, here is a system message: {foo()}")  # error, function does not return a value
```

It's a great one to save accidentally using something that represents no value.

`object` on the other hand could be anything, it could be a string, it has many methods itself.

Many other languages support this pattern because it's typesafe and a good convenience.

**Comment by erictraut** [comment]
I just realized I had typo in my response. I've fixed it above. Sorry if that caused confusion. Let me try again.

In the Python runtime there's no difference between a function that explicitly returns None and a function that implicitly returns None. They both have a return type of None. A `None` return type does not mean "void". It means that the function returns `None` either explicitly or implicitly. It's a sentinel value.

There is no equivalent of `void` in Python, either at runtime or in the type system.

If your intent is to describe a callable that takes a specified set of args but returns any type (i.e. you don't care about the return type), then you should use `Callable[<args>, Any]`.

In short, the current behavior is well defined and (I think) correct. The change you're proposing would cause a lot of breakage.

**Comment by KotlinIsland** [comment]
@erictraut 
```py
def foo() -> None
print(foo() is None)  # error, function does not return a value
```
A sole `None` return value actually behaves as `void` here, even though at runtime it is `None`, the type system prevents and usages of it, effectively `void`.

**Comment by KotlinIsland** [comment]
Just as in TypeScript a type of `void` can be represented with any value:
```ts
const fn: () => void = () => 100
console.log(fn())  // prints 100
```

**Comment by erictraut** [comment]
That doesn't cause a runtime error, nor is it an error in pyright.

It appears that mypy does generate an error here. I don't fully agree with that behavior, as it is inconsistent and doesn't match the runtime behavior. Furthermore, there's nothing in PEP 484 or any other type standards that would suggest this is the intended behavior of a `None` return type.

Yes, I understand that TypeScript has a dedicated `void` type with its own semantics. Python does not. We could add one if we think there's a need for it, but I don't think that need is very compelling. We should not overload `None` to have that meaning because `None` has its own well-defined semantics.

**Comment by KotlinIsland** [comment]
Kotlin takes it one step further:
```kt
val fn: (arg: Int) -> Unit = { it }  // no error
println(fn(1))  // prints "kotlin.Unit"
```

**Comment by KotlinIsland** [comment]
Hmm, didn't realize this was non-standard behavior, IMO the semantics of `None` are a mess already, as it refers to the literal value of `None` but also the type of `None`. 
eg:
```py
a: None = None
b: type[None] = type(None)  # no error
c: type[Literal[True]] = type(True)  # error
```

**Comment by erictraut** [comment]
Does `Callable[<args>, Any]` not meet your needs?

```python
fn: Callable[[int], Any] = lambda arg: arg
```

**Comment by KotlinIsland** [comment]
Okay then, if this is already non-standard mypy behavior I'll raise it in mypy.

**Comment by KotlinIsland** [comment]
> Does `Callable[<args>, Any]` not meet your needs?
> 
> ```python
> fn: Callable[[int], Any] = lambda arg: arg
> ```

No, that returns `Any`

**Comment by erictraut** [comment]
It's a function that returns a value you don't care about. Isn't that what you're trying to describe?

**Comment by KotlinIsland** [comment]
```py
from typing import Any, Callable
fn: Callable[[int], Any] = lambda arg: arg  # error: Explicit "Any" is not allowed
print(f"hi {fn(1).AMONGUS}")  # error: Expression has type "Any"
```
Not really what I want, I want a function that returns `void`, which is not part of the standardized types. I think `object` is the best official method of doing this.

**Comment by ethanhs** [comment]
Yeah we shouldn't complain when comparing to `None` like that, please open a mypy issue @KotlinIsland

**Comment by KotlinIsland** [comment]
> Yeah we shouldn't complain when comparing to `None` like that, please open a mypy issue @KotlinIsland

I love the fact that mypy has `void` like behavior. I don't want that code to be valid.
